excellent/types/base.go

```diff
@@ -0,0 +1,262 @@
+func (x XBool) ToString() XString { return XString(strconv.FormatBool(x.Native())) }
+
+// ToBool converts this type to a bool
+func (x XBool) ToBool() XBool { return x }
+
+// ToJSON converts this type to JSON
+func (x XBool) ToJSON() XString { return RequireMarshalToXString(x.Native()) }
+
+// Native returns the native value of this type
+func (x XBool) Native() bool { return bool(x) }
+
+var NilXBool = NewXBool(false)
+var _ XPrimitive = NilXBool
+
+// XTime is a point in time
+type XTime time.Time
+
+// NewXTime creates a new XTime
+func NewXTime(value time.Time) XTime {
+	return XTime(value)
+}
+
+// Type is the XType of this type
+func (x XTime) Type() XType { return XTypeTime }
+
+// ToString converts this type to a string
+func (x XTime) ToString() XString { return XString(utils.DateToISO(x.Native())) }
+
+// ToBool converts this type to a bool
+func (x XTime) ToBool() XBool { return XBool(!x.Native().IsZero()) }
+
+// ToJSON converts this type to JSON
+func (x XTime) ToJSON() XString { return RequireMarshalToXString(utils.DateToISO(x.Native())) }
+
+// Native returns the native value of this type
+func (x XTime) Native() time.Time { return time.Time(x) }
+
+var NilXTime = NewXTime(time.Time{})
+var _ XPrimitive = NilXTime
+
+// XError is an error
+type XError interface {
+	XPrimitive
+	error
+}
+
+type xerror struct {
+	err error
+}
+
+// NewXError creates a new XError
+func NewXError(err error) XError {
+	return xerror{err: err}
+}
+
+// Type is the XType of this type
+func (x xerror) Type() XType { return XTypeError }
+
+// ToString converts this type to a string
+func (x xerror) ToString() XString { return XString(x.Native().Error()) }
+
+// ToBool converts this type to a bool
+func (x xerror) ToBool() XBool { return XBool(false) }
+
+// ToJSON converts this type to JSON
+func (x xerror) ToJSON() XString { return RequireMarshalToXString(x.Native().Error()) }
+
+// Native returns the native value of this type
+func (x xerror) Native() error { return x.err }
+
+func (x xerror) Error() string { return x.err.Error() }
+
+var NilXError = NewXError(nil)
+var _ XError = NilXError
+
+// XObject is the interface for any complex object in Excellent
+type XObject interface {
+	XValue
+
+	Reduce() XPrimitive
+}
+
+// BaseXObject is base of any XObject
+type BaseXObject struct{}
+
+// Type is the XType of this type
+func (x *BaseXObject) Type() XType { return XTypeObject }
+
+// RequireMarshalToXString calls json.Marshal in the given value and panics in the case of an error
+func RequireMarshalToXString(x interface{}) XString {
+	j, err := json.Marshal(x)
+	if err != nil {
+		panic(fmt.Sprintf("unable to marshal %v to JSON", x))
+	}
+	return XString(j)
+}
+
+// ToXString converts the given value to a string
+func ToXString(value XValue) XString {
+	switch x := value.(type) {
+	case XPrimitive:
+		return x.ToString()
+	case XObject:
+		return x.Reduce().ToString()
+	}
+	panic(fmt.Sprintf("can't convert type %v to a string", value))
```

Comment:
using panics here because if those goes wrong we've coded something wrong - i.e. its more serious than giving the user back an `XError`